keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,423 @@
+      * [Alternatives](#alternatives)
+
+## Summary
+
+This proposal aims at allowing Pod resource requests & limits to be updated
+in-place, without a need to restart the Pod or its Containers.
+
+The **core idea** behind the proposal is to make PodSpec mutable with regards to
+Resources, denoting **desired** resources.
+Additionally, PodStatus is extended to provide information about **actual**
+resource allocation.
+
+This document builds upon [proposal for live and in-place vertical scaling][] and
+[Vertical Resources Scaling in Kubernetes][].
+
+[proposal for live and in-place vertical scaling]: https://github.com/kubernetes/community/pull/1719
+[Vertical Resources Scaling in Kubernetes]: https://docs.google.com/document/d/18K-bl1EVsmJ04xeRq9o_vfY2GDgek6B6wmLjXw-kos4/edit?ts=5b96bf40
+
+## Motivation
+
+Resources allocated to a Pod's Container can require a change for various reasons:
+* load handled by the Pod has increased significantly and current resources are
+  not sufficient,
+* load has decreased significantly and allocated resources are unused and wasted,
+* resources have simply been set improperly.
+
+Currently, changing resource allocation requires the Pod to be recreated since
+the PodSpec's Container Resources is immutable.
+
+While many stateless workloads are designed to withstand such a disruption, some
+are more sensitive, especially when using low number of Pod replicas.
+
+Moreover, for stateful or batch workloads, a Pod restart is a serious
+disruption, resulting in lower availability or higher cost of running.
+
+Allowing Resources to be changed without recreating the Pod or restarting the
+Containers addresses this issue directly.
+
+### Goals
+
+* Primary: allow to change Pod resource requests & limits without restarting its
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource resize is not possible.
+* Secondary: allow users to specify which Pods and Containers can be resized
+  without a restart.
+
+### Non-Goals
+
+The explicit non-goal of this KEP is to avoid controlling full life-cycle of a
+Pod which failed in-place resource resizing. This should be handled by actors
+which initiated the resizing.
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart,
+* updating extended resources or any other resource types besides CPU, memory.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to Container resources requests and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodSpec is extended with information about actually allocated Container resources.
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration, denoting
+  **desired** state of the Pod resources,
+* PodSpec.Container.ResourceAllocations (new object) denotes **actual** resources
```

Comment:
Curious why do we need ResourceAllocations field in PodSpec now ? I mean what requirements of this proposal mandate its introduction ? Also why not put it in pod status ?

Reply:
Yes, this is intentional.  The status should be an observation of the actual state of a pod, not a declaration of desired state.  Pod status should be able to be completely reconstructed based on the state of a node.  The kubelet should only respond to changes in the desired state of a pod, aka Spec.

Reply:
> @dashpole , why PodSpec and no longer PodStatus, was this an intentional move?

Yes. This change came from discussions with @derekwaynecarr and @dashpole about the [design principle](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status) that PodStatus should not store state, and if it were lost, can be regenerated on the fly. In the previous iteration, I ended up relying on state stored in PodStatus.

Tracking ResourceAllocations in PodSpec (in addition to ContainerStatus.ResourcesAllocated - which is purely reporting at this point) allows Scheduler to track the approved resources in a place expected to be persistent, and if Scheduler were to restart, it can arrive at where things were prior to restart without relying on Status.

Reply:
Well, the idea was that PodStatus was used to show the actual state while PodSpec was about the desired state - so that's in line with the design principle. I hope that we never suggested to use PodStatus as the desired state, if it was somehow implied I'd rather fix that single part than move the status into PodSpec.

As for kubelet responding only to change in desired state - as far as I understand the controllers (kubelet included) drive the actual state towards desired one. I'd assume that this usually requires knowing the actual state thus you could say that a controller acts on the actual state too, if only to check that no action is needed.

As for regenerating the status - it would be doable based on cgroups, right? The cited design princple says: "At some cost and perhaps some temporary degradation in behavior, the status could be reconstructed by observation if it were lost." so I don't see a problem here.

I'm a little afraid of putting actual state into PodSpec - to me it seems contrary to the convention where spec is only the desired state.

Or have I misunderstood something?

Reply:
The previous iteration required the kubelet to have pod status because it needed to verify that the scheduler had "approved" the new resource requirements.

It looks like we might have a misunderstanding.  The `ResourceAllocations` field in the pod spec is written by the scheduler, and doesn't represent the actual state of anything.  After the scheduler writes the field, the kubelet picks up the update, and reconciles that desired state by changing cgroup values.  The kubelet then observes the new cgroup values by inspecting the state of the node, and writes those into `ResourcesAllocated` in the pod status.

Just as before, `ResourcesAllocated` in the pod status represents the actual state of the cgroups  as observed by the kubelet.  The change is that instead of using a pod condition to denote that the scheduler has "approved" the new desired state, we explicitly separate the desired state of the scheduler (`ResourceRequirements`), and the desired state of the kubelet (`ResourcesAllocated`).

Reply:
interesting, so the new thing we are introducing is what is schedulers decision of what the pod can get in the podspec. It would help for the readers to document this clearly. Also requests from user is per container but scheduler decision is being recorded per podspec. It would also help to define why we do this way

Reply:
> interesting, so the new thing we are introducing is what is schedulers decision of what the pod can get in the podspec. It would help for the readers to document this clearly. Also requests from user is per container but scheduler decision is being recorded per podspec. It would also help to define why we do this way

Today, scheduler operates at pod resource request i.e sum(container-resource-requests) atomicity - so when it assigns a pod to a node, all containers get what they asked for. This design retains that property by making scheduler be aware of changes to resource request. I'll call this out explicitly in the KEP.

Reply:
> I've been trying to think about ways to avoid overloading the terms "Allocation" (used both for scheduler and kubelet actions), and "Resize" (used in both conditions). In the spirit of my comment above, where I suggest re-using the "Binding" pod subresource, maybe we can refer to the scheduler's action as "Binding", and the kubelet's action as "Allocating" or "Resizing". We could call the new field in the spec something like `PodSpec.Container.ResourcesBound`, and call the condition something like `PodResourcesBound` to mirror `PodScheduled`.

This terminology is better. I'm adopting it with minor changes. I'll push and update to the KEP by tonight or tomorrow morning. Thanks.

Reply:
Scheduler is not in the loop anymore. PodSpec.Containers[].ResourceAllocations now tracks the resources Kubelet agrees to. The updated design obsoletes this discussion. Resolving this issue.